packages/transport/src/api/usb.ts

```diff
@@ -119,32 +120,50 @@ export class UsbApi extends AbstractApi {
         }));
     }
 
-    private abortableMethod<R>(method: () => R, signal?: AbortSignal) {
+    private abortableMethod<R>(
+        method: () => R,
+        { signal, onAbort }: { signal?: AbortSignal; onAbort?: () => void },
+    ) {
         if (!signal) {
             return method();
         }
         if (signal.aborted) {
             return Promise.reject(new Error(ERRORS.ABORTED_BY_SIGNAL));
         }
 
-        return Promise.race([
-            // todo: maybe we need to pass signal down the chain?
-            method(),
-            new Promise<R>((_, reject) => {
-                signal?.addEventListener('abort', () => {
-                    reject(new Error(ERRORS.ABORTED_BY_SIGNAL));
-                });
-            }),
-        ]);
+        const abort = () => {
+            if (!onAbort) return;
+            try {
+                onAbort();
+            } catch {
+                // silent
+            }
+        };
+
+        const dfd = createDeferred<R>();
+        const abortListener = () => {
+            abort();
+            dfd.reject(new Error(ERRORS.ABORTED_BY_SIGNAL));
+        };
+        signal?.addEventListener('abort', abortListener);
+
+        return Promise.race([method(), dfd.promise])
```

Comment:
note for me: this is probably place to serialize all promises to run in sequence. will explain why later